vaadin-select-flow/src/main/java/com/vaadin/flow/component/select/Select.java

```diff
@@ -755,8 +765,22 @@ private void reset() {
         if (isEmptySelectionAllowed()) {
             addEmptySelectionItem();
         }
-        getDataProvider().fetch(new Query<>()).map(this::createItem)
-                .forEach(this::add);
+
+        synchronized (dataProvider) {
+            final AtomicInteger itemCounter = new AtomicInteger(0);
+            getDataProvider().fetch(new Query<>()).map(this::createItem)
+                    .forEach(component -> {
+                        add(component);
+                        itemCounter.incrementAndGet();
+                    });
+            lastFetchedDataSize = itemCounter.get();
+        }
+
+        runBeforeClientResponse(ui -> {
+            // Size event is fired before client response so as to avoid
+            // multiple size change events during server round trips
+            fireSizeEvent();
+        });
```

Comment:
```suggestion
        if(sizeRequest == null) {
            // Size event is fired before client response so as to avoid
            // multiple size change events during server round trips
            sizeRequest = context -> [
                fireSizeEvent();
                sizeRequest = null;
            }
            runBeforeClientResponse(ui ->sizeRequest());
        }
```